### PR TITLE
MUI Theming

### PR DESCRIPTION
**Pull request by jacob6838**
# PR Details

## Description

Enabling theming of the webapp through MUI. This involved the following changes:
- converting color references to MUI theme references
- converting some UI elements to MUI elements (mostly dropdowns and buttons)
- Creating styled components to replace some (not all) css formatting
- Creating 2 new themes: 'light' and 'dark', alongside slightly modified 'cdotDark' theme (this was the original theme)
- Allowing the icon to be set through environment variables, for light and dark modes
    - reading active theme mode (light/dark)
- Allowing the theme to be set through environment variables, for light and dark modes
    - reading browser light/dark mode

**NOTE: This PR is built off of the keycloak-custom-user-provider PR**

## How Has This Been Tested?

This has been extensively tested locally, by comparing updated elements to previous UI for cdotDark theme, and manually judging contrast/visual quality for new themes

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [x] My changes require new environment variables:
  - [x] I have updated the docker-compose, K8s YAML, and all dependent deployment configuration files.
- [x] My changes require updates to the documentation:
  - [x] I have updated the documentation accordingly.
- [x] My changes require updates and/or additions to the unit tests:
  - [x] I have modified/added tests to cover my changes.
- [x] All existing tests pass.

### CODE REVIEW

**Comment by drewjj** [comment]
Looks like some merge conflicts exist here.

**Comment by payneBrandon** [comment]
@jacob6838 looks like one of the unit tests is failing

**Comment by jacob6838** [comment]
Sample theme screenshots:
CDOT Dark:
cvmanager_cdot_dark_admin_2
cvmanager_cdot_dark_map_2
Dark:
cvmanager_main_dark_admin_2
cvmanager_main_dark_map_2
Light:
cvmanager_main_light_admin_2
cvmanager_main_light_map_2
Current Develop:
cvmanager_develop_admin
cvmanager_develop_map

File: webapp/src/styles/components/BorderedImage.tsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { styled } from '@mui/material'
+
+// display: block;
```
Comment (payneBrandon) [inline webapp/src/styles/components/BorderedImage.tsx]: Cleanup?

File: webapp/src/styles/components/HorizontalTabs.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import { styled } from '@mui/material'
+import { Link } from 'react-router-dom'
+
+// font-family: Arial, Helvetica, sans-serif;
```
Comment (payneBrandon) [inline webapp/src/styles/components/HorizontalTabs.tsx]: Can we remove the unused css?

File: webapp/src/pages/Map.tsx [hunk 3]
```diff
@@ -656,91 +753,87 @@ function MapPage(props: MapPageProps) {
...
+                  <FormGroup row>
+                    <Button
+                      variant="contained"
+                      className="contained-button"
+                      sx={{
+                        borderRadius: 4,
+                        width: '100%',
+                        '&.Mui-disabled': {
+                          backgroundColor: alpha(theme.palette.primary.light, 0.5),
+                          //   color: theme.palette.getContrastText(theme.palette.primary.light),
```
Comment (payneBrandon) [inline webapp/src/pages/Map.tsx]: Cleanup?

File: webapp/src/pages/Map.tsx [hunk 4]
```diff
@@ -994,19 +1084,30 @@ function MapPage(props: MapPageProps) {
...
                 id="stepSelect"
-                defaultValue={stepValueToOption(filterStep)}
-                placeholder={stepValueToOption(filterStep)}
-                onChange={(e) => dispatch(setGeoMsgFilterStep(e.value))}
-                options={stepOptions}
-              />
-              <button className="searchButton" onClick={() => dispatch(setGeoMsgFilter(false))}>
+                onChange={(e) => dispatch(setGeoMsgFilterStep(Number(e.target.value)))}
+                value={stepValueToOption(filterStep)?.value?.toString()}
+                // options={stepOptions}
```
Comment (payneBrandon) [inline webapp/src/pages/Map.tsx]: Cleanup?

**Comment by payneBrandon** [comment]
Overall changes are looking good, but there seems to be quite a bit of left over commented code that we could tidy up

File: webapp/src/pages/404.tsx [hunk 5]
```diff
@@ -76,7 +73,7 @@ export const NotFound = ({
       ) : (
         <></>
       )}
-      <h1 style={{ color: 'white' }}></h1>
+      <h1></h1>
```
Comment (payneBrandon) [inline webapp/src/pages/404.tsx]: Does this get injected with something, or is it stylized to just have a bar or something? Confused why it's in here

File: webapp/src/features/intersections/notifications/notifications-table.tsx [hunk 6]
```diff
@@ -269,15 +269,14 @@ export const NotificationsTable = (props: { simple: Boolean }) => {
...
+            <Grid2 container justifyContent="left" spacing={3}>
+              <Grid2>
                 <Button
                   sx={{
                     m: 1,
-                    color: 'white', // Normal state text color
-                    '&.Mui-disabled': {
-                      color: 'grey', // Disabled state text color
-                    },
+                    // '&.Mui-disabled': {
```
Comment (payneBrandon) [inline webapp/src/features/intersections/notifications/notifications-table.tsx]: Cleanup?

**Review by payneBrandon — CHANGES_REQUESTED** [review]
Overall looking good, just a few questions

File: webapp/src/styles/components/BorderedImage.tsx (hunk 1 above)
Comment (jacob6838) [reply webapp/src/styles/components/BorderedImage.tsx]: updated

File: webapp/src/styles/components/HorizontalTabs.tsx (hunk 2 above)
Comment (jacob6838) [reply webapp/src/styles/components/HorizontalTabs.tsx]: updated

File: webapp/src/pages/Map.tsx (hunk 3 above)
Comment (jacob6838) [reply webapp/src/pages/Map.tsx]: updated

File: webapp/src/pages/Map.tsx (hunk 4 above)
Comment (jacob6838) [reply webapp/src/pages/Map.tsx]: updated

File: webapp/src/features/intersections/notifications/notifications-table.tsx (hunk 6 above)
Comment (jacob6838) [reply webapp/src/features/intersections/notifications/notifications-table.tsx]: updated

File: webapp/src/pages/404.tsx (hunk 5 above)
Comment (jacob6838) [reply webapp/src/pages/404.tsx]: removed!

**Comment by payneBrandon** [comment]
Looks like one of the unit tests is failing

File: webapp/src/utils/test-utils.tsx [hunk 7]
```diff
@@ -32,4 +32,21 @@ function replaceChaoticIds(container: HTMLElement) {
   return container
 }
 
-export { replaceChaoticIds }
+import React from 'react'
```
Comment (payneBrandon) [inline webapp/src/utils/test-utils.tsx]: To match our other files should the imports go to the top?

File: webapp/src/setupTests.ts [hunk 8]
```diff
@@ -27,14 +27,29 @@ Object.defineProperty(window, 'localStorage', {
   value: localStorageMock,
 })
 
+// const matchMediaMock = function (query) {
```
Comment (payneBrandon) [inline webapp/src/setupTests.ts]: Needed?

File: sample.env [hunk 9]
```diff
@@ -105,8 +105,10 @@ MAPBOX_INIT_ZOOM="10"
 
 # Webapp themes: light, dark, cdotDark
 # WEBAPP_THEME_LIGHT value is used if the browser is in light mode. If the browser is in dark mode, WEBAPP_THEME_DARK is used
-WEBAPP_THEME_LIGHT="dark" # if not set, defaults to 'light'
-# WEBAPP_THEME_DARK="dark" # if not set, defaults to 'dark'
+# if not set, defaults to 'light'
+WEBAPP_THEME_LIGHT="cdotDark"
```
Comment (payneBrandon) [inline sample.env:109]: Did you say this actually needs to be REACT_APP_WEBAPP_THEME_LIGHT to get it to pop?

Comment (jacob6838) [reply sample.env:109]: That is only for the npm-specific .env file (.env.local)

File: webapp/src/utils/test-utils.tsx (hunk 7 above)
Comment (jacob6838) [reply webapp/src/utils/test-utils.tsx]: updated!

File: webapp/src/setupTests.ts (hunk 8 above)
Comment (jacob6838) [reply webapp/src/setupTests.ts]: removed

File: webapp/src/components/tables/GenericTable.tsx [hunk 10]
```diff
@@ -0,0 +1,142 @@
+// import { SvgIconComponent } from '@mui/icons-material'
```
Comment (mwodahl) [inline webapp/src/components/tables/GenericTable.tsx]: Is this just intended to serve as a reference file?

**Review by mwodahl — COMMENTED** [review]
This all looks good - I just had one question.

I did notice when running locally that the keycloak login theme doesn't change with these new updates - is that something that we're planning on updating to fit the theme?

I also wanted to note that it looks like we're using the lighter orange color again for buttons in the cdotDark theme. I believe this passes the WCAG AA requirements as we're only using large text on these buttons, however, we won't be able to use smaller text on these going forward as it'll fail that requirement.

**Comment by jacob6838** [comment]
@mwodahl 

The other big thing that changed was that I am not using the white-backed admin tables... what is your opinion on this? it required another separate theme, which is why I omitted it, but I think can add it back with custom theme props, just adds a few more colors per theme... what do you think?

**Comment by jacob6838** [comment]
> I did notice when running locally that the keycloak login theme doesn't change with these new updates - is that something that we're planning on updating to fit the theme?

@mwodahl I can create a task for this, but that doesn't seem necessary to me at this time. The example keycloak theme doesn't have any icons, the "Sign In" button background color is all that would change

File: webapp/src/components/tables/GenericTable.tsx (hunk 10 above)
Comment (jacob6838) [reply webapp/src/components/tables/GenericTable.tsx]: removed!

**Comment by mwodahl** [comment]
> @mwodahl
> 
> The other big thing that changed was that I am not using the white-backed admin tables... what is your opinion on this? it required another separate theme, which is why I omitted it, but I think can add it back with custom theme props, just adds a few more colors per theme... what do you think?

I think it's fine without the white-backed admin tables - I don't see any contrast violations (although they could occur if the user implements an additional theme). If you'd like to re-add them I think that would be fine as well. I liked the white-backed admin tables because they add more contrast, however, if it would require another separate theme I don't think they're necessary.

**Review by mwodahl — APPROVED** [review]
This all looks good to me - unit tests are passing and each of the themes worked well when run locally!

File: docker-compose-full-cm.yml [hunk 11]
```diff
@@ -69,8 +69,13 @@ services:
         MAPBOX_INIT_ZOOM: ${MAPBOX_INIT_ZOOM}
         CVIZ_API_SERVER_URL: ${CVIZ_API_SERVER_URL}
         CVIZ_API_WS_URL: ${CVIZ_API_WS_URL}
+        WEBAPP_THEME_LIGHT: ${WEBAPP_THEME_LIGHT}
```
Comment (drewjj) [inline docker-compose-full-cm.yml:72]: Please make a pull request in CDOT DevOps for new environment variables if these are required so we don't break the CDOT deployment once this is merged.

**Review by drewjj — APPROVED** [review]
Looks great! Just the one comment.

File: docker-compose-full-cm.yml (hunk 11 above)
Comment (jacob6838) [reply docker-compose-full-cm.yml:72]: These have been added to the devops pipeline, and there is an active PR to include the cdot icon in webapp deployments as a volume: https://dev.azure.com/SOC-OIT/CDOT/_git/cdot-cv-gcp/pullrequest/85984

**Review by payneBrandon — APPROVED** [review]
lgtm! unit tests pass and theme seems to be pulling in

**Pull request merged**